### PR TITLE
Add Speed augmentation on CPU, Time Stretch augmentation on CPU+GPU

### PR DESCRIPTION
**Pull request by titu1994**
# Changelog
## Additions
- SpeedPerturbation + TimeStretchPerturbation for sample level augmentation on the CPU
- TimeStretchAugmentation neural module for batch level augmentation on the GPU.

## Fixes
- Use double type to compute accuracy to correct bug on PyTorch 1.5

## Issues
- **GPU TimeStretchAugmentation can cause deadlock on Pytorch 1.4 and torchaudio 0.4**. This issue seems to be resolved in Pytorch 1.5 and torchaudio 0.5.
Should we check library version and inform user of possible deadlock?

### CODE REVIEW

**Comment by blisc** [comment]
It seems like torchaudio has stopped pinning torch versions. We should probably add them to our requirements now, and remove our torchaudio check.

For now, yes please raise a logging.error if torchaudio and pytorch versions are known to cause deadlocking issues.

File: nemo/collections/asr/audio_preprocessing.py [hunk 1]
```diff
@@ -738,6 +742,163 @@ def output_ports(self):
...
+
+        if probability > 1.0 or probability < 0.0:
+            raise ValueError("`probability` must be between 0 and 1")
+
+        if not HAVE_TORCHAUDIO:
+            raise ModuleNotFoundError(
+                "torchaudio is not installed but is necessary for "
+                "TimeStretchAugmentation. We recommend you try "
+                "installing it from conda for the PyTorch version you have."
+            )
```
Comment (blisc) [inline nemo/collections/asr/audio_preprocessing.py:792]: Let's add torchaudio to nemo_asr requirements, and remove these checks.

File: nemo/collections/asr/parts/perturb.py [hunk 2]
```diff
@@ -10,6 +10,13 @@
 from nemo.collections.asr.parts import collections, parsers
 from nemo.collections.asr.parts.segment import AudioSegment
 
+try:
+    from nemo.collections.asr.parts import numba_utils
+
+    HAVE_NUMBA = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_NUMBA = False
+
```
Comment (blisc) [inline nemo/collections/asr/parts/perturb.py:19]: If numba is easy to install, just add it to requirements and remove check

File: nemo/collections/asr/parts/perturb.py [hunk 3]
```diff
@@ -20,20 +27,164 @@ def perturb(self, data):
...
+            speed_rate = self._rng.uniform(self._min_rate, self._max_rate)
+        else:
+            speed_rate = self._rng.choice(self._rates)
+
+        # Skip perturbation in case of identity speed rate
+        if speed_rate == 1.0:
+            return
+
         if speed_rate <= 0:
             raise ValueError("speed_rate should be greater than zero.")
```
Comment (blisc) [inline nemo/collections/asr/parts/perturb.py:79]: Let's move this check to init

File: nemo/collections/asr/parts/perturb.py [hunk 4]
```diff
@@ -20,20 +27,164 @@ def perturb(self, data):
...
-        # logging.debug("speed: %f", speed_rate)
-        data._samples = librosa.effects.time_stretch(data._samples, speed_rate)
+
+        new_sr = int(self._sr * speed_rate)
+        data._samples = librosa.core.resample(data._samples, self._sr, new_sr, res_type='kaiser_fast')
```
Comment (blisc) [inline nemo/collections/asr/parts/perturb.py:82]: I would hesitate to default to kasier_fast, this would result in degraded ASR training.

File: nemo/collections/asr/audio_preprocessing.py [hunk 5]
```diff
@@ -738,6 +742,163 @@ def output_ports(self):
...
+        self,
+        sample_rate: int,
+        probability: float,
+        min_speed_rate: float = 0.9,
+        max_speed_rate: float = 1.1,
+        num_rates: int = 5,
+        n_fft: int = 512,
+    ):
+        """
+        Time-stretch an audio series by a fixed rate while preserving pitch.
```
Comment (blisc) [inline nemo/collections/asr/audio_preprocessing.py:756]: This time-stretches a batch of audio signals not just 1 audio signal. I also assume that all elements in the batch is stretched by the same amount?

File: tests/unit/test_unit_speech_commands.py [hunk 6]
```diff
@@ -86,6 +86,10 @@ def test_pytorch_audio_dataset_with_perturbation(self):
         perturbations = [
             perturb.WhiteNoisePerturbation(min_level=-90, max_level=-46),
             perturb.ShiftPerturbation(min_shift_ms=-5.0, max_shift_ms=5.0),
+            perturb.TimeStretchPerturbation(min_speed_rate=0.9, max_speed_rate=1.1),
+            perturb.SpeedPerturbation(
+                sr=self.featurizer_config['sample_rate'], min_speed_rate=0.9, max_speed_rate=1.1
+            ),
```
Comment (blisc) [inline tests/unit/test_unit_speech_commands.py:92]: Can you test each perturbation separately and add asserts that check they do what they are supposed to?

**Comment by blisc** [comment]
Thank you for adding a changelog comment.

File: nemo/collections/asr/audio_preprocessing.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/audio_preprocessing.py:756]: Yes it performs one time stretch (randomly sampled) on all items of that batch. I'll make it clear in the docstring.

File: nemo/collections/asr/audio_preprocessing.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/audio_preprocessing.py:792]: Sounds good.

File: nemo/collections/asr/parts/perturb.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:19]: Numba is particularly hard to install using pip. On Windows, it doesn't work cause of llvmlite package. Conda is the preferred method of installing it, which is why I am doing an optional check.

File: nemo/collections/asr/parts/perturb.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:79]: Good point.

File: nemo/collections/asr/parts/perturb.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:82]: I was thinking of speed over quality. I'll expose it as a parameter, and default to Kaiser fast. Is that fine?

Kaiser best runtime cost is 2.7x on average that of Kaiser fast, which already takes over 100 ms per file of maximum duration 16.7 seconds in my speed tests.

File: tests/unit/test_unit_speech_commands.py (hunk 6 above)
Comment (titu1994) [reply tests/unit/test_unit_speech_commands.py:92]: Hmm for shift I can probably do a strider check. For white noise, do I just check the two items are not exactly the same Element wise ?

File: nemo/collections/asr/audio_preprocessing.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/audio_preprocessing.py:792]: Looking at torchaudio's installation procedure, it seems they recommend conda installs as well. The pip install version (https://github.com/pytorch/audio) fails on Windows, but the conda install works. I think its best to not explicitly add the pip version to requirements_asr.txt so as not to break windows installations.

File: nemo/collections/asr/parts/perturb.py (hunk 4 above)
Comment (blisc) [reply nemo/collections/asr/parts/perturb.py:82]: This is part of the reason we tend to shy away from doing online speed perturbation.
I'm not too sure about defaulting to kaiser_fast.

File: tests/unit/test_unit_speech_commands.py (hunk 6 above)
Comment (blisc) [reply tests/unit/test_unit_speech_commands.py:92]: I am not too sure. Maybe do element-wise perturbed - unperturbed. And then fit a guassian on it? And see if the mean and std is what you expect?

File: nemo/collections/asr/audio_preprocessing.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/asr/audio_preprocessing.py:792]: I wouldn't be too concerned about windows, but can you raise this issue on slack?

File: nemo/collections/asr/parts/perturb.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:82]: So with kaiser fast, I could increase the range of augmentations from [90%, 100%, 110%] to [90%, 95%, 100%, 105%, 110%] with a 2x slowdown of training. If we augment the data offline, the training time is technically 3x normal time. So there is some benefit to on the fly augmentation, even if its marginal. 

As to kaiser_fast vs kaiser_best, its a toss up. Personally id go for speed over quality, but making it a user choice is probably the best. Depending on amount of data and quality of original data, either option might be better.

File: nemo/collections/asr/audio_preprocessing.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/audio_preprocessing.py:792]: Hmm, even for linux and macosx, torchaudio has the `--find-links https://download.pytorch.org/whl/torch_stable.html` line in its installation, which has links to compiled binaries. That line has to be called at pip level, and cant be added into the requirements_asr.txt file itself. 

In my opinion, it is probably best to inform users that certain features will be disabled unless they install torchaudio separately as is currently done for certain pre-processing steps like MFCC.

File: nemo/collections/asr/parts/perturb.py (hunk 4 above)
Comment (blisc) [reply nemo/collections/asr/parts/perturb.py:82]: How about for ASR?

Let's just make a required argument.

Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py:82]: Sounds good.

File: tests/unit/test_unit_speech_commands.py (hunk 6 above)
Comment (titu1994) [reply tests/unit/test_unit_speech_commands.py]: Added individual checks for each of the augmentations.

File: nemo/collections/asr/parts/perturb.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py]: Made it a required arg

File: nemo/collections/asr/parts/perturb.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/parts/perturb.py]: Fixed.

File: nemo/collections/asr/audio_preprocessing.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/audio_preprocessing.py]: Added note to docstring

File: nemo/collections/asr/parts/numba_utils.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
```
Comment (blisc) [inline nemo/collections/asr/parts/numba_utils.py:1]: Please add a copyright header here

Comment (titu1994) [reply nemo/collections/asr/parts/numba_utils.py:1]: Oh forgot about that one. On it.

Comment (titu1994) [reply nemo/collections/asr/parts/numba_utils.py:14]: Fixed.
